File: SensoroSDKActive-ObjC/SensoroSDKActive-ObjC/SENLocationManager.m
//
//  SENLocationManager.m
//  BeaconActive-ObjC
//
//  Created by skyming on 15/4/10.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

#import "SENLocationManager.h"
#import <UIKit/UIKit.h>

@interface SENLocationManager ()<SBKBeaconManagerDelegate>

@end

@implementation SENLocationManager

+ (instancetype)sharedInstance
{
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init{
    
    if(self = [super init])
    {
        _locationManager = [SBKBeaconManager sharedInstance];

        if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 8.0)
        {
            //获取授权
            [_locationManager requestAlwaysAuthorization];
        }
        //设定代理
        _locationManager.delegate = self;
        //生成监测用的区域ID
        _monitorRegion = [SBKBeaconID beaconIDWithProximityUUID:[[NSUUID alloc]initWithUUIDString:@"46D06053-9FAD-483B-B704-E576735CE1A3"]];
        
    }
    return self;
}


- (void)startMonitor:(BOOL)relaunch{
    
    if (!relaunch) {
        [_locationManager startRangingBeaconsWithID:_monitorRegion wakeUpApplication:YES];
        NSLog(@"Start monitor region!");
    }else{
        NSLog(@"During the relauch app, don't restart monitor region!");

    }
    _started = true;

}

- (void)stopMonitor{
    [_locationManager stopRangingBeaconsWithID:_monitorRegion];
    _started = false;
    NSLog(@"Stop monitor region!");
}

- (void)sendNotification:(NSString *)notify{
    UILocalNotification *notification = [[UILocalNotification alloc]init];
    notification.alertBody = notify;
    notification.soundName = UILocalNotificationDefaultSoundName;
    
    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
}


- (void)beaconManager:(SBKBeaconManager *)beaconManager didRangeNewBeacon:(SBKBeacon *)beacon{
   
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    NSLog(@"Enter: %@",now);
}

- (void)beaconManager:(SBKBeaconManager *)beaconManager beaconDidGone:(SBKBeacon *)beacon{
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    NSLog(@"Exit: %@",now);
}

- (void)beaconManager:(SBKBeaconManager *)beaconManager didDetermineState:(SBKRegionState)state forRegion:(SBKBeaconID *)region{
  
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    switch (state) {
        case CLRegionStateInside:
        {
            NSString *alertbody =[NSString stringWithFormat:@"Enter region at %@",[formatter stringFromDate:now]];
            [self sendNotification:alertbody];
            NSLog(@"%@",alertbody);
        }
            break;
            
        case CLRegionStateOutside:
        {
            NSString *alertbody =[NSString stringWithFormat:@"Exit region at %@",[formatter stringFromDate:now]];
            [self sendNotification:alertbody];
            NSLog(@"%@",alertbody);
        }
            break;
            
        case CLRegionStateUnknown:
            NSLog(@"This region state was unknown!");
            break;
    }

}

@end

File: SensoroSDKActive-iOS/SensoroSDKActive-iOS/SENLocationManager.swift
//
//  SENLocationManager.swift
//  SensoroSDKActive-iOS
//
//  Created by David Yang on 15/4/8.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

import UIKit
import CoreLocation
import SensoroBeaconKit

private let _sharedInstance = SENLocationManager();

class SENLocationManager: NSObject, SBKBeaconManagerDelegate {
    
    class var sharedInstance : SENLocationManager {
        return _sharedInstance;
    }

    var started = false;
    var monitorRegion : SBKBeaconID!;

    
    private override init(){
        
        super.init();

        //设定代理
        SBKBeaconManager.sharedInstance().delegate = self;
        //获取授权
        SBKBeaconManager.sharedInstance().requestAlwaysAuthorization();
        //生成监测用的区域ID
        monitorRegion = SBKBeaconID(proximityUUID: NSUUID(UUIDString: "46D06053-9FAD-483B-B704-E576735CE1A3"));
    }
    
    func startMonitor( relaunch : Bool ){
        if relaunch == false {
            SBKBeaconManager.sharedInstance().startRangingBeaconsWithID(monitorRegion, wakeUpApplication: true);
            NSLog("Start monitor region!");
        }else{
            NSLog("During the relauch app, don't restart monitor region!");
        }
        started = true;
    }
    
    
    func stopMonitor(){
        SBKBeaconManager.sharedInstance().stopRangingBeaconsWithID(monitorRegion);
        started = false;
        NSLog("Stop monitor region!");
    }
    
    func sendNotification(notify : String){
        var notification = UILocalNotification()
        notification.alertBody = notify
        
        notification.soundName = UILocalNotificationDefaultSoundName
        UIApplication.sharedApplication().scheduleLocalNotification(notification)
    }
    
    // MARK: SBKBeaconManagerDelegate
    func beaconManager(beaconManager: SBKBeaconManager!, didRangeNewBeacon beacon: SBKBeacon!) {
        let now = NSDate();
        let formatter = NSDateFormatter();
        formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
        //NSLog("Enter a beacon at \(formatter.stringFromDate(now))");
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, beaconDidGone beacon: SBKBeacon!) {
        let now = NSDate();
        let formatter = NSDateFormatter();
        formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
        //NSLog("Leave a region at \(formatter.stringFromDate(now))");
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, scanDidFinishWithBeacons beacons: [AnyObject]!) {
        //NSLog("Periodically call for scanDidFinishWithBeacons");
    }
    
    func beaconManager(beaconManager: SBKBeaconManager!, didDetermineState state : SBKRegionState, forRegion region : SBKBeaconID!) {
        
        switch state {
        case .Enter:
            let now = NSDate();
            let formatter = NSDateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification("Enter region at \(formatter.stringFromDate(now))");
            NSLog("Enter region at \(formatter.stringFromDate(now))");
        case .Leave:
            let now = NSDate();
            let formatter = NSDateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification("Exit  region at \(formatter.stringFromDate(now))");
            NSLog("Exit  region at \(formatter.stringFromDate(now))");
        case .Unknown:
            NSLog("This region state was unknown!");
        }
    }
}
